cEP-0026.md

```diff
@@ -0,0 +1,192 @@
+| Type     | Feature                                   |
+
+## Abstract
+
+This document describes how a caching mechanism (an enhanced version of the one
+being used by the old core) will be implemented and integrated with the
+NextGen-Core as a part of this
+[GSoC project](https://summerofcode.withgoogle.com/projects/#6434190552203264).
+
+## Introduction
+
+Since the NextGen-Core aims to provide faster speed and better performance for
+coala-runs on large codebases it needs caching to achieve this kind of
+performance.
+
+Structure wise the NextGen cache is a dictionary with bear types and cache
+tables as key-value pairs. The cache tables themselves are
+dictionary-like-objects that map the `persistent_hash`s of the task objects to
+the bear results. This current mechanism will be further enhanced by adding
+features like `FileFactory` class to load files, `Directory` class to interface
+with the directories, ignoring unmodified directories during successive runs and
+adding cache control flags for better usability and providing the user various
+options to perform caching according to a project's needs.
+
+## Implementation
+
+The implementation will be divided into four parts
+
+- Prototype for `FileFactory` class implementation.
+- Prototype for `Directory` class implementation.
+- Ignore directories functionality.
+- Cache control flags.
+
+### Prototype of FileFactory class implementation
+
+This class will follow the Factory design pattern and will act as a proxy to
+dealing with files. This class will have various methods to provide the file
+contents in different forms like string, binary format, etc. With the help of
+this class we can replace the actual file contents in our file-dict with
+these objects which will itself be a performance improvement.
+
+The following is the prototype for `FileFactory` class.
+
+```python
+import linecache
+import os
+
+from coala_utils.decorators import generate_eq
+from coalib.core.PersistentHash import persistent_hash
+
+
+@generate_eq('filename')
+class FileFactory:
+    def __init__(self, filename):
+        self.filename = filename
+        self.parent = os.path.abspath(os.path.join(self.filename, os.pardir))
+
+    @classmethod
+    def open(cls):
+        return cls(self.filename)
+
+    def file_as_bytes(self):
+        with open(self.filename, 'rb') as fp:
+            return fp.read()
+
+    def file_as_string(self):
+        return self.file_as_bytes().decode()
+
+    def file_as_list(self):
+        self.lines = linecache.getlines(self.filename)
+        self.linecount = len(self.lines)
+
+    def get_line(self, line):
+        return linecache.getline(self.filename, line)
+
+    @property
+    def timestamp(self):
+        # get the last modified timestamp of the file
+        return os.path.gettime(self.filename)
+
+    def hash(self):
+        # hash the absolute file path to get a unique hash
+        return persistent_hash(self.filename)
```

Comment:
We might wanna store it as data member depending on the complexity of this function call.
Although it can be taken care of while implementing. So no need to change it right now. Just keep this in mind.